tomo_challenge/jax_metrics.py

```diff
@@ -0,0 +1,229 @@
+    fractions = np.array([c / len(labels) for c in counts])
+    n_eff = np.array([n_eff_total_arcmin2 * f for f in fractions])
+    # Create redshift bins
+    nzs = []
+    for i in range(nbins):
+        if binned_nz:
+            # In this case, we use a histogram instead of a KDE, to make things a lot faster
+            h, he = onp.histogram(labels, bins=512, range=[0,4], weights=weights[:,i], density=True)
+            he = 0.5*(he[1:]+he[:-1])
+            nz = jc.redshift.kde_nz(he, h, bw=4./512, gals_per_arcmin2=n_eff[i], zmax=4.)
+        else:
+            nz = jc.redshift.kde_nz(labels, weights[:,i], bw=kernel_bandwidth,
+                                      gals_per_arcmin2=n_eff[i], zmax=4.)
+        nzs.append(nz)
+    probes = []
+    # start with number counts
+    if (what == 'gg' or what == '3x2'):
+        # Define a bias parameterization
+        bias = jc.bias.inverse_growth_linear_bias(1.)
+        probes.append(jc.probes.NumberCounts(nzs, bias))
+
+    if (what == 'ww' or what == '3x2'):
+        probes.append(jc.probes.WeakLensing(nzs, sigma_e=sigma_e))
+
+    return probes
+
+def compute_snr_score(weights, labels, what='3x2', binned_nz=False):
+    """Compute a score metric based on the total spectrum S/N
+
+    This is given by sqrt(mu^T . C^{-1} . mu) - baseline
+    where mu is the theory prediction and C the Gaussian covariance
+    for this set of bins. The baseline is the score for no tomographic binning.
+
+    Parameters
+    ----------
+    tomo_bin: array
+        Tomographic bin choice (0 .. bin_max) for each object in the survey
+    z: array
+        True redshift for each object
+
+    Returns
+    -------
+    score: float
+        Metric for this configuration
+    """
+    # Retrieve the probes
+    probes = get_probes(weights, labels, what=what, binned_nz=binned_nz)
+    ell, delta_ell = ell_binning()
+
+    @jax.jit
+    def snr_fn(probes, ell):
+        # instantiates fiducial cosmology
+        cosmo = jc.Cosmology(
+            Omega_c = 0.27,
+            Omega_b = 0.045,
+            h = 0.67,
+            n_s = 0.96,
+            sigma8 = 0.8404844953840714,
+            Omega_k=0.,
+            w0=-1., wa=0.)
+
+        # Compute mean and covariance
+        mu, C = jc.angular_cl.gaussian_cl_covariance_and_mean(cosmo, ell, probes,
+                                                              f_sky=0.25,
+                                                              nonlinear_fn=jc.power.halofit,
+                                                              sparse=True)
+
+        # S/N for correlated data, I assume, from generalizing
+        # sqrt(sum(mu**2/sigma**2))
+        P = jc.sparse.inv(C)
+        score = (mu.T @ jc.sparse.sparse_dot_vec(P, mu))**0.5
+        return score
+
+    return snr_fn(probes, ell)
+
+def compute_fom(weights, labels, inds=[0,4], what='3x2', binned_nz=False):
+    """
+    Computes the omega_c, sigma8 Figure of Merit
```

Comment:
Can JAX do the w0-wa one?  I always wanted that one, it was just a limitation of CCL that we used this one.

Reply:
Yes it can :-D As a matter of fact, if you specify "FOM_DETF_3x2" as the metric in the yaml file it will compute it